tests/integration/test_healing.py

```diff
@@ -0,0 +1,281 @@
+#!/usr/bin/env python3
+# Copyright 2022 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+import logging
+from pathlib import Path
+from time import sleep
+
+import pytest
+import yaml
+from helpers import (
+    app_name,
+    cluster_name,
+    cut_network_from_unit,
+    execute_commands_on_unit,
+    generate_random_string,
+    get_primary_unit,
+    get_process_pid,
+    get_system_user_password,
+    graceful_stop_server,
+    instance_ip,
+    is_connection_possible,
+    is_unit_in_cluster,
+    restore_network_for_unit,
+    scale_application,
+    start_server,
+    unit_hostname,
+    wait_network_restore,
+)
+from pytest_operator.plugin import OpsTest
+
+from src.constants import CLUSTER_ADMIN_USERNAME, SERVER_CONFIG_USERNAME
+
+logger = logging.getLogger(__name__)
+
+METADATA = yaml.safe_load(Path("./metadata.yaml").read_text())
+APP_NAME = METADATA["name"]
+MYSQL_DAEMON = "mysqld"
+
+
+async def build_and_deploy(ops_test: OpsTest) -> None:
+    """Build and deploy."""
+    if app := await app_name(ops_test):
+        if len(ops_test.model.applications[app].units) == 3:
+            return
+        else:
+            async with ops_test.fast_forward():
+                await scale_application(ops_test, app, 3)
+            return
+
+    # Build and deploy charm from local source folder
+    charm = await ops_test.build_charm(".")
+
+    # Reduce the update_status frequency until the cluster is deployed
+    async with ops_test.fast_forward():
+        await ops_test.model.deploy(charm, application_name=APP_NAME, num_units=3)
+
+        await ops_test.model.block_until(
+            lambda: len(ops_test.model.applications[APP_NAME].units) == 3
+        )
+        await ops_test.model.wait_for_idle(
+            apps=[APP_NAME],
+            status="active",
+            raise_on_blocked=True,
+            timeout=1000,
+        )
+
+
+@pytest.mark.order(1)
+@pytest.mark.abort_on_fail
+@pytest.mark.healing_tests
+async def test_kill_db_process(ops_test: OpsTest) -> None:
+    """Kill mysqld process and check for auto cluster recovery."""
+    await build_and_deploy(ops_test)
+
+    app = await app_name(ops_test)
+    unit = ops_test.model.applications[app].units[0]
+    another_unit = ops_test.model.applications[app].units[1]
+
+    # get running mysqld PID
+    pid = await get_process_pid(ops_test, unit.name, MYSQL_DAEMON)
+
+    # kill mysqld for the unit
+    logger.info(f"Killing process id {pid}")
+    await ops_test.juju("ssh", unit.name, "sudo", "kill", "-9", pid)
```

Comment:
I think you'll want to check the results of this to verify that kill was successful

Reply:
I do that by comparing PID's a bit further in the test, given that the mysqld daemon will not start if there's another daemon running.